docs/en/DevelopingUFSExtensions.md

```diff
@@ -0,0 +1,138 @@
+text file in `META_INF/services` with a single line pointing to the class implementing the said
+interface.
+
+## Dependency Management
+
+Implementors are required to include transitive dependencies of the extension in the built JAR. To
+avoid dependency conflicts from including dependencies of an extension project on the server
+classpath, the ServiceLoader in Alluxio uses a custom (per-extension)
+[ClassLoader](https://docs.oracle.com/javase/7/docs/api/java/lang/ClassLoader.html) for isolation.
+
+# Implementing an Under Storage Extension
+
+Building a new under storage connector involves: 
+
+- Implementing the required under storage interface
+- Declaring the service implementation
+- Bundling up the implementation and transitive dependencies in an uber JAR
+
+A reference implementation can be found in the [alluxio-extensions](https://github.com/Alluxio
+/alluxio-extensions/tree/master/underfs/template) repository. In the rest of this section we
+describe the steps involved in writing a new under storage extension. The sample project, called
+`DummyUnderFileSystem`, uses maven as the build and dependency management tool, and forwards all
+operations to a local filesystem.
+
+## Implement the Under Storage Interface
+
+Refer to [integrating under storage systems](Integrating-Under-Storage-Systems.html) for
+additional instructions for developing and testing an under storage module.
+
+Step 1: Implement the interface `UnderFileSystem`
+
+The `UnderFileSystem` interface is defined in the module `org.alluxio:alluxio-core-common`. Choose
+to extend either `BaseUnderFileSystem` or `ObjectUnderFileSystem` to implement the `UnderFileSystem`
+interface. `ObjectUnderFileSystem` is suitable for connecting to object storage and abstracts away
+mapping file system operations to an object store.
+
+```java
+public class DummyUnderFileSystem extends BaseUnderFileSystem {
+	// Implement filesystem operations
+	...
+}
+```
+
+or,
+
+```java
+public class DummyUnderFileSystem extends ObjectUnderFileSystem {
+	// Implement object store operations 
+	...
+}
+```
+
+Step 2: Implement the interface `UnderFileSystemFactory`
+
+The under storage factory determines when an interface of the implemented `UnderFileSystem` is
+instantiated.
+
+```java
+public class DummyUnderFileSystemFactory implements UnderFileSystemFactory {
+	...
+	public boolean supportsPath(String path) {
+		// Choose which schemes to support, e.g., dummy://
+	}
+}
+```
+
+## Declare the Service
+
+Create a file at `src/main/resources/META_INF/services/alluxio.underfs.UnderFileSystemFactory`
+advertising the implemented `UnderFileSystemFactory` to the ServiceLoader.
+
+```
+alluxio.underfs.dummy.DummyUnderFileSystemFactory
+```
+
+## Build
+
+Include all transitive dependencies of the extension project in the built JAR using either
+`maven-shade-plugin` or `maven-assembly`.
+
+In addition, to avoid collisions specify scope for the dependency `alluxio-core-common` as
```

Comment:
How does this help to avoid collisions?

Reply:
avoids definition of classes like alluxiouri using extension classloader